tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java

```diff
@@ -15,49 +15,35 @@
 
 package org.tensorflow;
 
-import java.util.HashMap;
-import java.util.Map;
-
-import org.tensorflow.types.UInt8;
-
-/** Represents the type of elements in a {@link Tensor} as an enum. */
-public enum DataType {
-  /** 32-bit single precision floating point. */
-  FLOAT(1, 4),
-
-  /** 64-bit double precision floating point. */
-  DOUBLE(2, 8),
-
-  /** 32-bit signed integer. */
-  INT32(3, 4),
-
-  /** 8-bit unsigned integer. */
-  UINT8(4, 1),
+import org.tensorflow.internal.c_api.TF_Tensor;
+import org.tensorflow.util.Shape;
+
+/** Represents a type of elements in a {@link Tensor} */
+public final class DataType<T> {
+
+  @FunctionalInterface
+  public interface TensorMapper<T> {
+
+    /**
+     * Maps tensor memory to a data structure for manipulating elements of this type.
+     *
+     * @param nativeTensor pointer to the native tensor
+     * @param shape the shape of the tensor
+     * @return data structure of elements of this type
+     */
+    T apply(TF_Tensor nativeTensor, Shape shape);
```

Comment:
This exposes the JavaCPP `TF_Tensor` type into the public API. Can we fix that?

Reply:
It is not "really" exposed. This interface is only used internally but it has to be public unfortunately because it is shared between packages. 

What I can do is to move it under `org.tensorflow.internal.buffer` so at least it is known to be internal. And since it does not return a `TF_Tensor` but accepts one, there is no major "leak" neither.

Reply:
It's times like this that I understand why C++ has friend classes.